File: domains/examples/polyglot/smart_lock/smart_lock.c
/* Deterministic contract lowered from the reviewed V2 domain 'smart_lock'.
 * Human review of the reviewed artifact is required before trust. */

typedef struct {
    int door_state;
    int lock_state;
} smart_lock;

/*@
  requires \valid(counter);
  assigns counter->door_state, counter->lock_state;
  ensures counter->door_state == 1 && counter->lock_state == 0;
  ensures ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
*/
void smart_lock_init(smart_lock *counter) {
    counter->door_state = 1;
    counter->lock_state = 0;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->door_state;
*/
int smart_lock_get_door_state(const smart_lock *counter) {
    return counter->door_state;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->lock_state;
*/
int smart_lock_get_lock_state(const smart_lock *counter) {
    return counter->lock_state;
}

/*@
  requires \valid(counter);
  requires ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
  requires counter->door_state == 0;
  requires counter->lock_state == 0;
  assigns counter->door_state;
  ensures counter->door_state == 1;
  ensures ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
*/
void smart_lock_close_door(smart_lock *counter) {
    counter->door_state = 1;
}

/*@
  requires \valid(counter);
  requires ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
  requires counter->door_state == 1;
  requires counter->lock_state == 0;
  assigns counter->door_state;
  ensures counter->door_state == 0;
  ensures ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
*/
void smart_lock_open_door(smart_lock *counter) {
    counter->door_state = 0;
}

/*@
  requires \valid(counter);
  requires ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
  requires counter->door_state == 1;
  requires counter->lock_state == 0;
  assigns counter->lock_state;
  ensures counter->lock_state == 1;
  ensures ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
*/
void smart_lock_lock_door(smart_lock *counter) {
    counter->lock_state = 1;
}

/*@
  requires \valid(counter);
  requires ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
  requires counter->lock_state == 1;
  assigns counter->lock_state;
  ensures counter->lock_state == 0;
  ensures ((0 <= counter->door_state) && (counter->door_state <= 1)) && ((0 <= counter->lock_state) && (counter->lock_state <= 1)) && ((counter->lock_state == 1) ==> (counter->door_state == 1));
*/
void smart_lock_unlock_door(smart_lock *counter) {
    counter->lock_state = 0;
}

File: domains/examples/polyglot/smart_lock/SmartLock.rs
use prusti_contracts::*;

pub struct SmartLock {
    pub door_state: i32,
    pub lock_state: i32,
}

impl SmartLock {
    #[ensures(result.door_state == 1 && result.lock_state == 0 && ((0 <= result.door_state) && (result.door_state <= 1)) && ((0 <= result.lock_state) && (result.lock_state <= 1)) && ((result.lock_state == 1) ==> (result.door_state == 1)))]
    pub fn new() -> Self {
        Self { door_state: 1, lock_state: 0 }
    }

    #[pure]
    #[ensures(result == self.door_state)]
    pub fn get_door_state(&self) -> i32 {
        self.door_state
    }

    #[pure]
    #[ensures(result == self.lock_state)]
    pub fn get_lock_state(&self) -> i32 {
        self.lock_state
    }

    #[requires(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    #[requires(self.door_state == 0)]
    #[requires(self.lock_state == 0)]
    #[ensures(self.door_state == 1)]
    #[ensures(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    pub fn close_door(&mut self) {
        self.door_state = 1;
    }

    #[requires(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    #[requires(self.door_state == 1)]
    #[requires(self.lock_state == 0)]
    #[ensures(self.door_state == 0)]
    #[ensures(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    pub fn open_door(&mut self) {
        self.door_state = 0;
    }

    #[requires(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    #[requires(self.door_state == 1)]
    #[requires(self.lock_state == 0)]
    #[ensures(self.lock_state == 1)]
    #[ensures(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    pub fn lock_door(&mut self) {
        self.lock_state = 1;
    }

    #[requires(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    #[requires(self.lock_state == 1)]
    #[ensures(self.lock_state == 0)]
    #[ensures(((0 <= self.door_state) && (self.door_state <= 1)) && ((0 <= self.lock_state) && (self.lock_state <= 1)) && ((self.lock_state == 1) ==> (self.door_state == 1)))]
    pub fn unlock_door(&mut self) {
        self.lock_state = 0;
    }
}
